client/src/main/java/org/apache/uniffle/client/impl/ShuffleWriteClientImpl.java

```diff
@@ -192,12 +206,27 @@ private boolean sendShuffleDataAsync(
     return result;
   }
 
-  private void genServerToBlocks(ShuffleBlockInfo sbi, List<ShuffleServerInfo> serverList,
-      Map<ShuffleServerInfo, Map<Integer, Map<Integer, List<ShuffleBlockInfo>>>> serverToBlocks,
-      Map<ShuffleServerInfo, List<Long>> serverToBlockIds) {
+  void genServerToBlocks(ShuffleBlockInfo sbi,
+                         List<ShuffleServerInfo> serverList,
+                         int replicaNum,
+                         List<ShuffleServerInfo> excludeServers,
+                         Map<ShuffleServerInfo, Map<Integer, Map<Integer, List<ShuffleBlockInfo>>>> serverToBlocks,
+                         Map<ShuffleServerInfo, List<Long>> serverToBlockIds,
+                         boolean includeBlockList) {
+    if (replicaNum <= 0) {
+      return;
+    }
     int partitionId = sbi.getPartitionId();
     int shuffleId = sbi.getShuffleId();
+    int assignedNum = 0;
     for (ShuffleServerInfo ssi : serverList) {
+      if (!includeBlockList && replica > 1 && !shuffleServerBlockList.isEmpty()
```

Comment:
Will it allocate the servers which is less than `replicaNum` if exclude nodes are too many?

Reply:
Updated